File: tests/roles/py/backend.py
"""Backend role: a backend serving request types until stopped.

--serves REQUEST=RESPONSE maps each request type to the reply type;
--behaviour says what to do with the payload (upper, echo, drop, raise,
sleep:MS); --crash-after N exits with status 3 after N requests. The role
stops on SIGTERM or when its --drain-file appears; with --drain-seconds it
drains first (declines searches, keeps serving) for that long, and with
--drain-min-handled N it refuses to leave before N requests were served.
--exit-on-exception makes a handler exception end the process with status 4.
"""

import os
import sys
import time

from tests.roles.py import common
from tests.roles.py.common import emit, servers


class Backend(servers.BaseMultiplexerServer):
    """The backend; see the module docstring for what the options do."""

    def __init__(
        self,
        addresses,
        type,
        serves,
        behaviour,
        crash_after,
        memory_every=0,
        drain_seconds=0.0,
        drain_file=None,
        drain_min_handled=0,
        exit_on_exception=False,
    ):
        super().__init__(addresses, type=type)
        self.serves = serves
        self.behaviour = behaviour
        self.crash_after = crash_after
        self.memory_every = memory_every
        self.drain_seconds = drain_seconds
        self.drain_file = drain_file
        self.drain_min_handled = drain_min_handled
        self.exit_on_exception = exit_on_exception
        self.handled = 0

    def handle_message(self, mxmsg):
        """Report the request, then answer, drop, raise or crash as configured."""
        self.handled += 1
        if self.memory_every and self.handled % self.memory_every == 0:
            common.memory_event(self.handled)
        emit("request", type=mxmsg.type, id=mxmsg.id, from_=mxmsg.from_, size=len(mxmsg.message))
        response_type = self.serves.get(mxmsg.type)
        if response_type is None:
            emit("unexpected", type=mxmsg.type, id=mxmsg.id)
            self.no_response()
            return
        if self.behaviour == "drop":
            self.no_response()
        elif self.behaviour == "raise":
            raise RuntimeError("handler failed on purpose")
        else:
            if self.behaviour.startswith("sleep:"):
                time.sleep(int(self.behaviour.split(":", 1)[1]) / 1000.0)
            payload = mxmsg.message.upper() if self.behaviour == "upper" else mxmsg.message
            self.send_message(message=payload, type=response_type)
        if self.crash_after and self.handled >= self.crash_after:
            emit("crash", handled=self.handled)
            os._exit(3)

    def periodic_task(self):
        """Notice a request to leave, from SIGTERM or the drain file: stop at
        once, or start draining when a drain was configured."""
        if self.draining:
            return
        if common.STOP.is_set() or (self.drain_file and os.path.exists(self.drain_file)):
            if not self.drain_seconds and not self.drain_min_handled:
                self.working = False
                return
            self.start_draining()  # keep serving, but no longer answer searches
            emit("draining", drain_seconds=self.drain_seconds)

    def drained(self):
        """The drain period is over and at least --drain-min-handled requests were served."""
        return super().drained() and self.handled >= self.drain_min_handled

    def on_handler_exception(self, exc):
        """Keep serving, unless --exit-on-exception."""
        return not self.exit_on_exception


def main() -> None:
    """Parse the options and serve until asked to leave."""
    p = common.parser("multiplexer backend role")
    p.add_argument("--serves", action="append", help="REQUEST_TYPE=RESPONSE_TYPE, repeatable")
    p.add_argument("--behaviour", default="upper", help="upper | echo | drop | raise | sleep:MS")
    p.add_argument("--crash-after", type=int, default=0, help="exit(3) after N handled requests")
    p.add_argument("--memory-every", type=int, default=0, help="emit a memory event every N requests")
    p.add_argument(
        "--drain-seconds",
        type=float,
        default=0.0,
        help="when asked to leave, decline searches but keep serving this long, then exit",
    )
    p.add_argument("--drain-file", default=None, help="a file whose appearance asks the backend to leave")
    p.add_argument("--drain-min-handled", type=int, default=0, help="do not leave before N requests were served")
    p.add_argument("--exit-on-exception", action="store_true", help="a handler exception ends the process (4)")
    args = p.parse_args()
    common.stop_on_sigterm()

    common.start_memory_tracing(args.memory_every)
    backend = Backend(
        common.endpoints(args),
        args.type,
        common.kv_ints(args.serves),
        args.behaviour,
        args.crash_after,
        args.memory_every,
        args.drain_seconds,
        args.drain_file,
        args.drain_min_handled,
        args.exit_on_exception,
    )
    emit(
        "connected", instance_id=backend.conn.instance_id, connections=backend.conn.connections_count(), name=args.name
    )
    try:
        backend.serve_forever(poll=0.25, drain_seconds=args.drain_seconds)
    except Exception as exc:  # the handler's exception, let through by on_handler_exception
        emit("handler_exception", kind=common.error_name(exc), handled=backend.handled)
        sys.exit(4)
    emit("stopped", handled=backend.handled)


if __name__ == "__main__":
    main()

File: tests/roles/cc/backend.cc
// The `backend` role: a backend that serves request types, with knobs for
// the failure scenarios. See tests/README.md for options and events.
#include <chrono>
#include <cstdlib>
#include <stdexcept>
#include <thread>
#include <unistd.h>

#include "multiplexer/backend/base_multiplexer_server.h"
#include "mxcontrol/task.h"
#include "mxcontrol/tasks_holder.h"
#include "tests/roles/cc/common.h"

namespace mxtestroles {

using multiplexer::MultiplexerMessage;
using mx::util::kwargs::Kwargs;

// How the role is asked to leave and what it does then; see the module
// comment of the Python role for the options.
struct LeaveOptions {
  double drain_seconds = 0.0;
  std::string drain_file;
  int drain_min_handled = 0;
  bool exit_on_exception = false;
};

// The backend that answers requests. --serves maps each request type to the
// reply type; --behaviour says what to do with the payload: upper (the
// default), echo, drop (no reply), raise (throw from the handler), sleep:MS
// (wait, then upper-case). --crash-after N exits with status 3 after N
// requests, mid-run, for the failover scenarios. Events: request,
// unexpected, crash, draining, stopped.
class BackendServer : public multiplexer::backend::BaseMultiplexerServer {
public:
  BackendServer(Client *client, unsigned type, const std::map<boost::uint32_t, boost::uint32_t> &serves,
                const std::string &behaviour, int crash_after, int memory_every, const LeaveOptions &leave)
      : BaseMultiplexerServer(client, type), serves_(serves), behaviour_(behaviour), crash_after_(crash_after),
        memory_every_(memory_every), leave_(leave), handled_(0) {}

  int handled() const { return handled_; }

protected:
  // Report the request, then answer, drop, raise or crash as configured.
  void handle_message(MultiplexerMessage &mxmsg) override {
    ++handled_;
    if (memory_every_ && handled_ % memory_every_ == 0)
      emit(memory_event(handled_));
    Event request = event("request");
    request.set_type(mxmsg.type());
    request.set_id(mxmsg.id());
    request.set_from_(mxmsg.from());
    request.set_size(mxmsg.message().size());
    emit(request);

    std::map<boost::uint32_t, boost::uint32_t>::const_iterator served = serves_.find(mxmsg.type());
    if (served == serves_.end()) {
      Event unexpected = event("unexpected");
      unexpected.set_type(mxmsg.type());
      unexpected.set_id(mxmsg.id());
      emit(unexpected);
      no_response();
      return;
    }
    if (behaviour_ == "drop") {
      no_response();
    } else if (behaviour_ == "raise") {
      throw std::runtime_error("handler failed on purpose");
    } else {
      if (behaviour_.compare(0, 6, "sleep:") == 0)
        std::this_thread::sleep_for(std::chrono::milliseconds(boost::lexical_cast<int>(behaviour_.substr(6))));
      std::string payload = behaviour_ == "upper" ? upper(mxmsg.message()) : mxmsg.message();
      send_message(Kwargs().set("message", payload).set("type", static_cast<boost::uint32_t>(served->second)));
    }
    if (crash_after_ && handled_ >= crash_after_) {
      Event crash = event("crash");
      crash.set_handled(handled_);
      emit(crash);
      std::_Exit(3);
    }
  }

  // Notice a request to leave, from SIGTERM or the drain file: stop at
  // once, or start draining when a drain was configured.
  void periodic_task() override {
    if (draining())
      return;
    bool asked = stop_requested || (!leave_.drain_file.empty() && access(leave_.drain_file.c_str(), F_OK) == 0);
    if (!asked)
      return;
    if (leave_.drain_seconds <= 0 && leave_.drain_min_handled <= 0) {
      working = false;
      return;
    }
    start_draining(); // keep serving, but no longer answer searches
    Event draining_event = event("draining");
    draining_event.set_drain_seconds(leave_.drain_seconds);
    emit(draining_event);
  }

  // The drain period is over and at least --drain-min-handled requests were served.
  bool drained() const override { return BaseMultiplexerServer::drained() && handled_ >= leave_.drain_min_handled; }

  // Keep serving, unless --exit-on-exception.
  bool on_handler_exception(const std::exception &) override { return !leave_.exit_on_exception; }

private:
  std::map<boost::uint32_t, boost::uint32_t> serves_;
  std::string behaviour_;
  int crash_after_;
  int memory_every_;
  LeaveOptions leave_;
  int handled_;
};

// The `backend` subcommand: runs a BackendServer until asked to leave.
class BackendRole : public mxcontrol::Task {
public:
  virtual std::string short_description() const { return "serve request types until stopped"; }
  virtual int run() {
    install_signal_handlers();
    std::unique_ptr<Client> client = common_.connect();
    BackendServer server(client.get(), common_.type, kv_ints(serves_), behaviour_, crash_after_, memory_every_, leave_);
    emit(common_.connected_event(*client));
    try {
      server.serve_forever(0.25f, static_cast<float>(leave_.drain_seconds));
    } catch (std::exception &error) { // the handler's exception, let through by on_handler_exception
      Event failed = event("handler_exception");
      failed.set_kind(error.what());
      failed.set_handled(server.handled());
      emit(failed);
      return 4;
    }
    Event stopped = event("stopped");
    stopped.set_handled(server.handled());
    emit(stopped);
    return 0;
  }

protected:
  virtual void _initialize_options_description(po::options_description &options) {
    common_.add(options);
    options.add_options()("serves", po::value(&serves_)->composing(), "REQUEST_TYPE=RESPONSE_TYPE, repeatable")(
        "behaviour", po::value(&behaviour_)->default_value("upper"), "upper | echo | drop | raise | sleep:MS")(
        "crash-after", po::value(&crash_after_)->default_value(0), "exit(3) after N handled requests")(
        "memory-every", po::value(&memory_every_)->default_value(0), "emit a memory event every N requests")(
        "drain-seconds", po::value(&leave_.drain_seconds)->default_value(0.0),
        "when asked to leave, decline searches but keep serving this long, then exit")(
        "drain-file", po::value(&leave_.drain_file)->default_value(""),
        "a file whose appearance asks the backend to leave")("drain-min-handled",
                                                             po::value(&leave_.drain_min_handled)->default_value(0),
                                                             "do not leave before N requests were served")(
        "exit-on-exception", po::bool_switch(&leave_.exit_on_exception), "a handler exception ends the process (4)");
  }

private:
  CommonOptions common_;
  std::vector<std::string> serves_;
  std::string behaviour_;
  int crash_after_;
  int memory_every_;
  LeaveOptions leave_;
};

REGISTER_MXCONTROL_SUBCOMMAND(backend, BackendRole);

} // namespace mxtestroles
